### ISSUE TITLE
Steps towards a real tooptip implementation

### ISSUE THREAD

**Issue by cpitclaudel** [company-mode/company-mode#471]
Hi Dmitry,

We've discussed using a real popup instead of overlays a number of times, so I thought I'd investigate a bit. It seems that it should be possible to use `x-pos-tip` on graphical terminals (I think you mentioned that on another thread, too). Here's what it would look like:

```
(setq x-gtk-use-system-tooltips nil)

(defvar my/lines
  '(#(" combine-after-change-calls        " 0 1 (face (company-tooltip-selection company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common-selection company-tooltip-selection company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip-selection company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-fg))
    #(" combine-after-change-execute      " 0 1 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-bg))
    #(" combine-and-quote-strings         " 0 1 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-bg))
    #(" comint--complete-file-name-data   " 0 1 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-bg))
    #(" comint--match-partial-filename    " 0 1 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-bg))
    #(" comint--requote-argument          " 0 1 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-bg))
    #(" comint--unquote&expand-filename   " 0 1 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-bg))
    #(" comint--unquote&requote-argument  " 0 1 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-bg))
    #(" comint--unquote-argument          " 0 1 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-bg))
    #(" comint-accumulate                 " 0 1 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 1 4 (face (company-tooltip-common company-tooltip) mouse-face (company-tooltip-mouse)) 4 34 (face (company-tooltip) mouse-face (company-tooltip-mouse)) 34 35 (face company-scrollbar-bg))))

(x-show-tip (mapconcat (lambda (l) (concat l "​")) my/lines "\n") nil
            `((internal-border-width . 0)
              (border-width . 0)
              (background-color . ,(face-attribute 'company-tooltip :background))))
```

It looks pretty good, I think :)

Is this a direction that we want to explore? If so, is there a way for us to implement this little by little? I don't have much time at all, but I'd be very happy to see this happen.

Cheers,
Clément.

**Comment by stardiviner** [comment]
This looks nice.

**Comment by dgutov** [comment]
> Is this a direction that we want to explore?

I believe so. Switching the implementation to use it instead of overlays doesn't seem like it'll be the main difficulty. `x-show-tip` has drawbacks:

* Have you tried to show two tips side-by-side? If we can't, the user will have to choose between proper tooltips and `company-quickhelp`.

* The tooltip stays on top of all application windows, even after the user switches away from Emacs.

As far as the roadmap goes, it should be like this:

* Extract the tooltip-showing code to a separate package (a la popup.el).

* Implement different rendering backends within that package.

**Comment by cpitclaudel** [comment]
> Have you tried to show two tips side-by-side? If we can't, the user will have to choose between proper tooltips and company-quickhelp.

It's not possible; the C code specifically checks that there's at most one tooltip being displayed. The limiting factor here is that I don't know how to create a borderless frame. I asked here, to no avail: https://emacs.stackexchange.com/questions/20167/how-do-i-create-a-borderless-frame .

> The tooltip stays on top of all application windows, even after the user switches away from Emacs.

We can hide it from `focus-out-hook`, probably.

> Extract the tooltip-showing code to a separate package (a la popup.el).

Can we use pos-tip? I don't think it's in ELPA though.
Or where you thinking of a company-popup package?

**Comment by cpitclaudel** [comment]
What main difficulties do you envision, except for the incompatibility with company-quickhelp? (then again, maybe there's a way to get a borderless frame from Elisp; I'll ask on emacs-devel)

**Comment by dgutov** [comment]
> It's not possible; the C code specifically checks that there's at most one tooltip being displayed.

You'd have to start with `M-x report-emacs-bug`, I imagine. Then we'd support this method only with next Emacs, of course, but after that's polished we can consider enabling it for the users of the current releases, on demand.

> We can hide it from focus-out-hook, probably.

If that works on Alt-Tab, great.

> Can we use pos-tip?

I don't think so.

> Or where you thinking of a company-popup package?

Yes. Probably with a different name. And an API that's closer to popup.el than pos-tip.

**Comment by dgutov** [comment]
> What main difficulties do you envision

Off the top of my head, just those two. But we might notice further problems as we actively try it out.

**Comment by cpitclaudel** [comment]
>> It's not possible; the C code specifically checks that there's at most one tooltip being displayed.

> You'd have to start with M-x report-emacs-bug, I imagine. Then we'd support this method only with next Emacs, of course, but after that's polished we can consider enabling it for the users of the current releases, on demand.

I still have a small hope of finding a way to display multiple tooltips from Elisp. Then we could distribute this for current releases as well.

**Comment by dgutov** [comment]
A popup package extraction aside, if you just want to experiment, you could probably get away with just advising `company-pseudo-tooltip-show` and `company-pseudo-tooltip-hide`.

**Comment by dgutov** [comment]
> I still have a small hope of finding a way to display multiple tooltips from Elisp.

Sure. Eli should be the best person to discuss it with.

**Comment by cpitclaudel** [comment]
> A popup package extraction aside, if you just want to experiment, you could probably get away with just advising company-pseudo-tooltip-show and company-pseudo-tooltip-hide.

Ah, I'll have to learn this advising thing :)

**Comment by cpitclaudel** [comment]
Ok. I wrote to Emacs-devel: https://lists.gnu.org/archive/html/emacs-devel/2016-02/msg00713.html

**Comment by cpitclaudel** [comment]
I experimented a bit more with this. Unless I missed something, `x-show-tip` won't be enough, because we can't keep the tool-tip open and change its contents; which in turns means that the tool-tip flickers (disappears and pops back in) every time the selected line in the popup changes.

**Comment by dgutov** [comment]
Right. :-(

If you don't manage to solve flickering yourself, it's another item for the wishlist. But it probably means the `x-show-tip` based solution might not be adequate as a replacement for the users of current Emacs. Still I think it's the direction to go in.

Another option, for the less graphically-inclined, is to use `popup-menu`. Here's a discussion that touched on this: http://lists.gnu.org/archive/html/emacs-devel/2014-08/msg00128.html Not sure if we can show annotations this way (using a different color, or right-aligned, at least), but if we can, that could be a decent baseline solution.

I think Eli also mentioned that it's possible to make the menus not hijack the input focus, and show/hide them from Elisp, but I can't find that email now. This would also be a necessary element to use them as our popup backend.

**Comment by cpitclaudel** [comment]
> Right. :-(

> If you don't manage to solve flickering yourself, it's another item for the wishlist. But it probably means the x-show-tip based solution might not be adequate as a replacement for the users of current Emacs. Still I think it's the direction to go in.

Indeed; I spend more time looking at it, and it doesn't look like there's anything we can do about it. So it'll be tricky to do anything for current Emacsen.

There are some good news, though: Chris Feng pointed out on emacs-devel that we can experiment with xcb, a lisp implementation of the X11 protocol:

```elisp
  (require 'xcb)

  (setq frame (make-frame '((visibility . nil))))

  (let ((window (string-to-number (frame-parameter frame 'outer-window-id)))
        (connection (xcb:connect-to-socket)))
    (xcb:+request connection
        (make-instance 'xcb:ChangeWindowAttributes
                       :window window
                       :value-mask xcb:CW:OverrideRedirect
                       :override-redirect 1))
    (xcb:flush connection)
    (xcb:disconnect connection))

  (make-frame-visible frame)
``` 

Based on this, I built a demo at https://github.com/cpitclaudel/company-tooltip-frontend. It uses xcb, but presumably once Emacs exposes that functionality we can switch to it.

Limitations:

* Needs xcb, so only works in X.
* Resizing the tooltip is very very slow; we need to force redisplay every time the frame is resized. No idea if that's something we can do anything about... Right now it only grows; it doesn't shrink.
* Positioning is a bit tricky; I'm not sure I got it right (it works for me without menu bars or tool bars)

Let me know what you think.

> Another option, for the less graphically-inclined, is to use popup-menu. Here's a discussion that touched on this: http://lists.gnu.org/archive/html/emacs-devel/2014-08/msg00128.html. Not sure if we can show annotations this way (using a different color, or right-aligned, at least), but if we can, that could be a decent baseline solution.

I'm not sure. It doesn't look that good on TTYs, and it really  doesn't look/feel like a completion popup at all on graphical terminals, I think :/ I don't think it can do color, either (can it? There's no `x-gtk-use-system-popups` that we could set to nil AFAIS)

> I think Eli also mentioned that it's possible to make the menus not hijack the input focus, and show/hide them from Elisp, but I can't find that email now. This would also be a necessary element to use them as our popup backend.

I looked for that in another context, but I couldn't find anything.

**Comment by cpitclaudel** [comment]
Another thing:

I had to implement add a timer to ignore `hide` frontend event, because company sends one before every show, and that made the popup flicker.

**Comment by dgutov** [comment]
> Based on this, I built a demo at https://github.com/cpitclaudel/company-tooltip-frontend. It uses xcb, but presumably once Emacs exposes that functionality we can switch to it.

It looks pretty broken here: the frame has chrome, it's displayed in the top-left corner of the desktop, with dimensions that are too large, it takes focus, and it does not get updated even if I switch back to the original frame, and change selection in the popup. The original frame still shows the overlay-based popup, is that intended?

And it leaves network connections open, which becomes apparent when I try to close Emacs.

> Right now it only grows; it doesn't shrink.

That sounds not ideal, but maybe it's better than having character of different sizes distort the popup. We should wait for a core developer's response, though.

> I'm not sure. It doesn't look that good on TTYs, and it really doesn't look/feel like a completion popup at all on graphical terminals, I think :/ I don't think it can do color, either

Well, it's a menu. I'm sure some of the users would find it handy anyway. As long as the display engine can show multiple-width characters in the terminal, as well support some other trickier features, overlay-based tooltip's positioning won't be ideal, and non-graphical Emacs has nothing that would correspond to chromeless frames. It does support frames, though, some maybe adding ones that span less than the whole terminal is not out of the question.

**Comment by cpitclaudel** [comment]
> It looks pretty broken here: the frame has chrome, it's displayed in the top-left corner of the desktop, with dimensions that are too large, it takes focus, and it does not get updated even if I switch back to the original frame, and change selection in the popup.

Did you restart Emacs after installing xcb? And are you on 25?

> The original frame still shows the overlay-based popup, is that intended?

Sure; just replacing `pseudo-tooltip-{show,hide}` was causing all sorts of failures, so I just added the advice instead of replacing. We can fix this once we have something that starts being remotely usable :)

> And it leaves network connections open, which becomes apparent when I try to close Emacs.

I didn't test for that, but it could be xcb not installing properly.

**Comment by dgutov** [comment]
> Did you restart Emacs after installing xcb? And are you on 25?

No. Yes. Restarted it now, the chrome is gone, but the position is still the top-left corner, the contents still look wrong, and don't get updated as I scroll through the list: http://i.imgur.com/CbVXL5v.png

> We can fix this once we have something that starts being remotely usable :)

Sure.

> I didn't test for that, but it could be xcb not installing properly.

It probably was. There were no orphan processes this time.

**Comment by rswgnu** [comment]
@cpitclaudel, have a look at child frames in Emacs 26 for borderless frames.  Also, some sample code and discussion of this is here, in issue #745.

**Comment by tumashu** [comment]
In emacs26, tooltip may fast if you set tooltip-reuse-hidden-frame to t
http://repo.or.cz/emacs.git/commit/80864c2a04597d31ba453c9af69d35b15c4e1e24

I have tested its speed with my package: pyim, and found that it has no  flicker and can
be used as a frontend,  someone can try it.

The show speed at the moment is:        minibuffer ~= child-frame > popup-el > tooltip >> tooltip(tooltip-reuse-hidden-frame is nil)

**Comment by ianpan870102** [comment]
I a bit lost in the thread -- so is there a way to use company childframe (so that fringes and line numbers don't get blacked out)? And how does setting tooltip-reuse-hidden-frame relate to this?

**Comment by dgutov** [comment]
@ianpan870102 Just try https://github.com/tumashu/company-posframe/. It has some problems, but works well for some people.

**Comment by ianpan870102** [comment]
I was just playing around with it for the past 10 minutes! The only downside I see is the flickers when one types further to narrow down the candidates. But I guess it's a compromise since the posframe have to be redrawn again and again. Perhaps in the future the performance could be better.

**Comment by dgutov** [comment]
Flickering is usually caused by something different than performance problems. Currently posframe behaves somewhat differently across the supported range of desktop environment. We're working on that )on emacs-devel). Let's hope we get around to fixing flickering next.

**Comment by ianpan870102** [comment]
> We're working on that )on emacs-devel). Let's hope we get around to fixing flickering next.

That's great news! Looking forward to it

**Comment by dgutov** [comment]
This was finally resolved in https://github.com/company-mode/company-mode/pull/1525, thanks all!

**Comment by ianyepan** [comment]
Awesome news @dgutov , reading the thread you linked, do I understand correctly that the childframe frontend would not work for the current stable TUI Emacs 30.2? You mentioned "`Works in TTY too.`" but it's under the "`Things that stayed the same`" section.

I recently came across the [popon](https://codeberg.org/akib/emacs-popon) package, which [corfu-terminal](https://codeberg.org/akib/emacs-corfu-terminal) leverages to simulate a functional pop-up tooltip in TUI given that corfu notoriously hard-depends on childframe support. Popon takes a different approach than company frontends and solved the problem of disappearing left fringe line numbers, which is a limitation that company-mode in TUI has. I am not using popon or corfu-terminal and can't comment on the elegance or performance of its solution, but I thought anyone who was interested enough to comment on this thread is probably also interested to take a look or try them out.

Last question: I hear childframe support has arrived to TUI Emacs starting with Emacs 31 - does this benefit company-mode's latest childframe support directly? Or would it require some involved additional development to get childframe working in TUI Emacs 31?

**Comment by dgutov** [comment]
> do I understand correctly that the childframe frontend would not work for the current stable TUI Emacs 30.2?

Right: the TUI support is new in Emacs 31.

> You mentioned "Works in TTY too." but it's under the "Things that stayed the same" section.

"The same" as the external package company-posframe, which I extracted the core implementation from and rewrote some pieces.

> Popon takes a different approach than company frontends and solved the problem of disappearing left fringe line numbers, which is a limitation that company-mode in TUI has.

Last I checked, it used the one-overlay-per-line approach, same as popup-el back in the day. Which resulted in better behavior with line numbers but had more of other kinds of issues (usually with having the popup's lines straight up vertically). I haven't really tested popon myself, to tell how affected it is by the same, if at all.

>  thought anyone who was interested enough to comment on this thread is probably also interested to take a look or try them out

Certainly, and comments welcome. I just have to note that popup is not in GNU ELPA either, like popup-el, which originally prohibited us from using it as a dependency (and discouraged me from trying to debug its issues at the same time).

> I hear childframe support has arrived to TUI Emacs starting with Emacs 31 - does this benefit company-mode's latest childframe support directly?

Yes, through `posframe` which is now a dependency as well (and it has a few lines of code necessary to work well on TUI).
